/connect without parameters crashes

Whoops! Brain went into tf mode and did /connect without parameters (default world connect in tfland). This got me:

panic: runtime error: index out of range [0] with length 0

goroutine 18 [running]:
github.com/makyo/stimmtausch/client.(*Client).listen(0xc00018f920)
        /root/parts/stimmtausch/build/client/client.go:136 +0x3b4
created by github.com/makyo/stimmtausch/client.New
        /root/parts/stimmtausch/build/client/client.go:164 +0x156      

Restarting and doing it right (thus having a connection already), then figuring out how to reproduce the crash got a connection error:

09:10:15 ERROR stimmtausch.ui ui.go:358 error setting up connection in ui: FIFO for connection fm_warp already exists, cowardly not continuing

--- a/client/client.go
+++ b/client/client.go
@@ -132,6 +132,10 @@ func (c *Client) listen() {
 		res := <-c.listener
 		switch res.Name {
 		case "connect", "c":
+			if len(res.Payload) == 0 {
+				log.Errorf("no world or server specified")
+				continue
+			}
 			res.Name = "_client:connect"
 			_, err := c.Connect(res.Payload[0])
 			res.Err = err

--- a/config/client.go
+++ b/config/client.go
@@ -82,5 +82,6 @@ type UI struct {
 			DisconnectedMore   string `yaml:"disconnected_more" toml:"disconnected_more"`     // TODO
 			DisconnectedActive string `yaml:"disconnected_active" toml:"disconnected_active"` // TODO
 		} `yaml:"send_title" toml:"send_title"`
+		ModalTitle string `yaml:"modal_title" toml:"modal_title"`
 	}
 }

--- a/config/default.go
+++ b/config/default.go
@@ -112,4 +112,5 @@ stimmtausch:
           disconnected_more: "mediumvioletred+underline"
           # Disconnected world (non-focused) with unread lines
           disconnected_active: "deeppink3"
+        modal_title: "bold+cyan"
 `

--- a/help/builtins.go
+++ b/help/builtins.go
@@ -7,6 +7,17 @@
 package help
 
 var HelpMessages = map[string]Help{
+	"help": Help{
+		Name:      "/help",
+		ShortDesc: "command help",
+		Synopsis: map[string]string{
+			"":          "show this help",
+			"<command>": "show help for <command> (without the `/`)",
+		},
+		Overview:    "Give help on an internal command",
+		Description: "Find out more information on topics in Stimmtausch by using the /help command. For instance, to learn about the `/log` command, type `/help log`.\n\n`Available topics:`\n\n{HELPTOPICS}",
+	},
+
 	"log": Help{
 		Name:      "/log",
 		ShortDesc: "connection logging",
@@ -25,14 +36,16 @@ var HelpMessages = map[string]Help{
 		Name:      "/fg",
 		ShortDesc: "bring world to the foreground",
 		Synopsis: map[string]string{
-			"":        "rotate to the next world to the right",
+			"":        "rotate to the next active world to the right (same as `/]`)",
 			">":       "rotate to the next world to the right",
 			"<":       "rotate to the next world to the left",
+			"]":       "rotate to the next active world to the right",
+			"[":       "rotate to the next active world to the left",
 			"<world>": "switch to the named world",
 		},
 		Overview:    "Command to control moving between worlds.",
 		Description: "Moving between worlds in Stimmtausch is accomplished with the /fg command. You can rotate between worlds by using the special world names > and <, otherwise you can specifi which world you would like to bring to the foreground.",
-		SeeAlso:     "/> (same as `/fg >`), /< (same as `/fg <`)",
+		SeeAlso:     "`/>` (same as `/fg >`), `/<` (same as `/fg <`), `/]` (same as `/fg ]`), `/[` (same as `/fg [`)",
 	},
 
 	"connect": Help{
@@ -45,7 +58,7 @@ var HelpMessages = map[string]Help{
 		},
 		Overview:    "Command to connect to worlds.",
 		Description: "Connecting to worlds in Stimmtausch is accomplished with the /connect command. You can connect to worlds named in your configuration files.", // Additionally, you can connect to servers named in your configuration without user information, or to a specified address and port. In each of the latter two cases, you will be given a temporary world name which you can use with other commands.
-		SeeAlso:     "/disconnect, /fg",
+		SeeAlso:     "`/disconnect`, `/fg`",
 	},
 
 	"disconnect": Help{
@@ -56,7 +69,7 @@ var HelpMessages = map[string]Help{
 		},
 		Overview:    "Command to disconnect from worlds.",
 		Description: "Disconnecting from worlds in Stimmtausch is accomplished with the /disconnect command. It accepts a world name.",
-		SeeAlso:     "/connect",
+		SeeAlso:     "`/connect`",
 	},
 
 	"quit": Help{

--- a/help/text.go
+++ b/help/text.go
@@ -8,15 +8,21 @@ package help
 
 import (
 	"fmt"
+	"regexp"
 	"strings"
 	"text/template"
 
 	ansi "github.com/makyo/ansigo"
 )
 
+var referRe = regexp.MustCompile("`([^`]+)`")
+var longestRe = regexp.MustCompile("\t([^\t]+)\t")
+var ansiRe = regexp.MustCompile("\x1b[^m]+m")
+
 var textFuncs = template.FuncMap{
-	"synopsis": textSynopsis,
-	"seeAlso":  textSeeAlso,
+	"synopsis":    textSynopsis,
+	"seeAlso":     textSeeAlso,
+	"description": textDescription,
 }
 var textTemplate = template.Must(template.New("text").Funcs(textFuncs).Parse(`{{"\x1b[1m"}}OVERVIEW{{"\x1b[0m"}}
 
@@ -28,25 +34,52 @@ var textTemplate = template.Must(template.New("text").Funcs(textFuncs).Parse(`{{
 
 {{"\x1b[1m"}}DESCRIPTION{{"\x1b[0m"}}
 
-    {{.Description}}{{seeAlso .SeeAlso}}`))
+    {{description .Description}}{{seeAlso .SeeAlso}}`))
 
 func textSynopsis(name string, synMap map[string]string) string {
 	lines := []string{}
 	for cmd, desc := range synMap {
 		if len(cmd) > 0 {
-			lines = append(lines, fmt.Sprintf("\t%s %s \t%s", ansi.MaybeApplyOneWithReset("bold", name), ansi.MaybeApplyOneWithReset("underline", cmd), desc))
+			lines = append(lines, fmt.Sprintf("    \t%s %s\t%s", ansi.MaybeApplyOneWithReset("bold", name), ansi.MaybeApplyOneWithReset("underline", cmd), refer(desc)))
 		} else {
-			lines = append(lines, fmt.Sprintf("\t%s \t%s", ansi.MaybeApplyOneWithReset("bold", name), desc))
+			lines = append(lines, fmt.Sprintf("    \t%s\t%s", ansi.MaybeApplyOneWithReset("bold", name), refer(desc)))
 		}
 	}
-	return strings.Join(lines, "\n    ")
+	return strings.Join(tabSubst(lines), "\n    ")
 }
 
 func textSeeAlso(seeAlso string) string {
 	if len(seeAlso) != 0 {
-		return fmt.Sprintf("\n\n%s\n\n    %s", ansi.MaybeApplyOneWithReset("bold", "SEE ALSO"), seeAlso)
+		return fmt.Sprintf("\n\n%s\n\n    %s", ansi.MaybeApplyOneWithReset("bold", "SEE ALSO"), refer(seeAlso))
 	}
 	return ""
+}
+
+func textDescription(description string) string {
+	topics := []string{}
+	for cmd, h := range HelpMessages {
+		topics = append(topics, fmt.Sprintf("    \t%s\t%s", ansi.MaybeApplyWithReset("cyan", cmd), h.ShortDesc))
+	}
+	return strings.Join(tabSubst(topics), "\n    ")
+}
+
+func refer(s string) string {
+	return string(referRe.ReplaceAll([]byte(s), []byte(ansi.MaybeApplyWithReset("cyan", "$1"))))
+}
+
+func tabSubst(lines []string) []string {
+	longest := 0
+	for _, s := range lines {
+		cmd := ansiRe.ReplaceAll(longestRe.FindAllSubmatch([]byte(s), -1)[0][1], []byte(""))
+		if len(cmd) > longest {
+			longest = len(cmd)
+		}
+	}
+	for i, s := range lines {
+		cmd := longestRe.FindAllSubmatch([]byte(s), -1)[0][1]
+		lines[i] = string(longestRe.ReplaceAll([]byte(s), []byte(fmt.Sprintf("$1 %s", strings.Repeat(" ", longest-len(ansiRe.ReplaceAll([]byte(cmd), []byte("")))+1)))))
+	}
+	return lines
 }
 
 func RenderText(h Help) string {

--- a/ui/keybindings.go
+++ b/ui/keybindings.go
@@ -161,6 +161,24 @@ func (t *tui) scrollDown(g *gotui.Gui, v *gotui.View) error {
 	return nil
 }
 
+func (t *tui) scrollModalUp(g *gotui.Gui, v *gotui.View) error {
+	_, y := v.Origin()
+	if y != 0 {
+		v.SetOrigin(0, y-1)
+	}
+	return nil
+}
+
+func (t *tui) scrollModalDown(g *gotui.Gui, v *gotui.View) error {
+	_, y := v.Origin()
+	_, maxY := v.Size()
+	lines := len(v.ViewBufferLines())
+	if y+maxY <= lines {
+		v.SetOrigin(0, y+1)
+	}
+	return nil
+}
+
 // redraw forces a rerender of the current view in order to ensure that everything is in order
 func (t *tui) redraw(g *gotui.Gui, v *gotui.View) error {
 	log.Debugf("redrawing")
@@ -197,6 +215,23 @@ func (t *tui) activeWorldRight(g *gotui.Gui, v *gotui.View) error {
 
 func (t *tui) activeWorldLeft(g *gotui.Gui, v *gotui.View) error {
 	go t.client.Env.Dispatch("[", "")
+	return nil
+}
+
+func (t *tui) closeModal(g *gotui.Gui, v *gotui.View) error {
+	if err := g.DeleteView("modal"); err != nil {
+		return err
+	}
+	if err := g.DeleteView("modalTitle"); err != nil {
+		return err
+	}
+	if err := g.DeleteView("modalHelp"); err != nil {
+		return err
+	}
+	go g.Update(func(gg *gotui.Gui) error {
+		_, err := g.SetCurrentView("send")
+		return err
+	})
 	return nil
 }
 
@@ -239,6 +274,15 @@ func (t *tui) keybindings(g *gotui.Gui) error {
 		return err
 	}
 	if err := g.SetKeybinding("send", gotui.KeyEnd, gotui.ModNone, t.end); err != nil {
+		return err
+	}
+	if err := g.SetKeybinding("modal", gotui.KeyEnter, gotui.ModNone, t.closeModal); err != nil {
+		return err
+	}
+	if err := g.SetKeybinding("modal", gotui.KeyArrowUp, gotui.ModNone, t.scrollModalUp); err != nil {
+		return err
+	}
+	if err := g.SetKeybinding("modal", gotui.KeyArrowDown, gotui.ModNone, t.scrollModalDown); err != nil {
 		return err
 	}
 	return nil

--- a/ui/ui.go
+++ b/ui/ui.go
@@ -386,7 +386,12 @@ func (t *tui) listen() {
 			go t.client.Env.DirectDispatch(res)
 		case "help":
 			// get the command text and tell the system to display it in a modal
-			cmd := res.Payload[0]
+			var cmd string
+			if len(res.Payload) == 0 {
+				cmd = "help"
+			} else {
+				cmd = res.Payload[0]
+			}
 			h, ok := help.HelpMessages[cmd]
 			if !ok {
 				log.Warningf("no help available for /%s", cmd)
@@ -413,12 +418,51 @@ func (t *tui) listen() {
 			res.Name = "_tui:showModal"
 			go t.client.Env.DirectDispatch(res)
 		case "_tui:showModal":
-			fmt.Fprintf(t.currView.buffer, "~~~~~ %s\n%s\n~~~~~\n", res.Payload[0], res.Payload[1])
+			t.createModal(res.Payload[0], res.Payload[1])
 		default:
 			log.Tracef("got unknown signal result %v", res)
 			continue
 		}
 	}
+}
+
+func (t *tui) createModal(title, content string) {
+	go t.g.Update(func(g *gotui.Gui) error {
+		maxX, maxY := g.Size()
+		if v, err := g.SetView("modal", 3, 3, maxX-4, maxY-6); err != nil {
+			if err != gotui.ErrUnknownView {
+				log.Warningf("unable to create modal view %+v", err)
+				return err
+			}
+			v.Frame = true
+			v.FrameFgColor = gotui.ColorCyan | gotui.AttrBold
+			v.Wrap = true
+			v.WordWrap = true
+			fmt.Fprint(v, content)
+		}
+		if v, err := g.SetView("modalTitle", 5, 2, len(title)+8, 4); err != nil {
+			if err != gotui.ErrUnknownView {
+				log.Warningf("unable to create modal view %+v", err)
+				return err
+			}
+			v.Frame = false
+			fmt.Fprintf(v, " %s ", ansi.MaybeApplyWithReset(t.client.Config.Client.UI.Colors.ModalTitle, title))
+		}
+		modalHelpText := " Scroll: ↑/↓ | Close: <Enter> "
+		if v, err := g.SetView("modalHelp", maxX-3-len(modalHelpText), maxY-7, maxX-6, maxY-5); err != nil {
+			if err != gotui.ErrUnknownView {
+				log.Warningf("unable to create modal view %+v", err)
+				return err
+			}
+			v.Frame = false
+			fmt.Fprint(v, ansi.MaybeApplyWithReset(t.client.Config.Client.UI.Colors.ModalTitle, modalHelpText))
+		}
+		if _, err := g.SetCurrentView("modal"); err != nil {
+			log.Warningf("unable to set current view %+v", err)
+			return err
+		}
+		return nil
+	})
 }
 
 // Run creates the UI and starts the mainloop.

